qiskit/opflow/primitive_ops/primitive_op.py

```diff
@@ -246,7 +246,11 @@ def assign_parameters(self, param_dict: dict) -> OperatorBase:
                 return ListOp([self.assign_parameters(param_dict) for param_dict in unrolled_dict])
             if self.coeff.parameters <= set(unrolled_dict.keys()):
                 binds = {param: unrolled_dict[param] for param in self.coeff.parameters}
-                param_value = float(self.coeff.bind(binds))
+                param_value = self.coeff.bind(binds)
+                if np.imag(complex(param_value)) == 0:
+                    param_value = float(param_value)
+                else:
+                    param_value = complex(param_value)
```

Comment:
This is logically a better approach now, I think, yeah.  This implementation isn't quite right, though - for example `float(1 + 0j)` will raise a `TypeError` even though it's meant to be valid.  I think it should be
```suggestion
                param_value = complex(self.coeff.bind(binds))
                if abs(param_value.imag) == 0:
                    param_value = param_value.real
```

Reply:
Right. Why do you need `abs` there?